writing/2019-10-05-Designing-testable-components.md

```diff
@@ -0,0 +1,471 @@
+store the result. As the requirements grow, the solution above quickly becomes
+quite verbose in practice.
+
+E.g:
+
+```haskell
+validateUser :: (UserId -> IO (Maybe User)) -> (UserName -> IO Bool) -> UserId -> IO Bool
+validateUser getUser validateUsername userId = do
+  userM <- getUser userId
+  maybe (pure False) (validateUsername . userName) userM
+```
+
+This example adds one function as argument, but what if you add a third? A
+fourth? You get the picture.
+
+We also don't want to write our code in `IO` - while useful of course, the
+surface area of possible effects is huge. From an effect perspective, we'd like
+to limit the power of each component. We'll get to this later in the article,
+but first let's focus on solving scalability of this initial approach.
+
+### Introducing the Handle pattern
+Instead of parameterizing the function with another function we can
+parameterize with a datatype containing a function. This is great for a number
+of reasons.
+
+1. We can group functions that operate in similar ways together (think OO
+   interface)
+2. We don't have to pass around all those functions, instead we pass one
+   datatype
+
+So how would this look?
+
+```haskell
+data Application =
+  { persistUser :: UserName -> Password -> IO UserId
+  , getUser :: UserId -> IO (Maybe User)
+  , validateUsername :: UserId -> IO Bool
+  , logLn :: Loggable a => a -> IO ()
+  }
+```
+
+Now we can simply pass that to `validateUser`:
+
+```haskell
+validateUser :: Application -> RequestBody -> UserId -> IO Bool
+validateUser app requestBody userId = do
+  userM <- (app & getUser) userId
+  maybe (pure False) (app & validateUsername $ userName) userM
+```
+
+But - when looking at this, you've probably seen an issue. `validateUsername`
+does not really fit in with the rest of the functions in `Application`. As a
+solution, we could nest the `Application` type. So let's redefine it:
+
+```haskell
+data Persistence =
+  { persistUser :: UserName -> Password -> IO UserId
+  , getUser :: UserId -> IO (Maybe User)
+  }
+
+data Application =
+  { persistence :: Persistence
+  , validateUsername :: UserId -> IO Bool
+  }
+```
+
+This gives us a bit more granularity, and a cleaner interface to work with.
+However, we still have a couple of issues:
+
+* We're running in `IO`
+* We have to manually pass around the `Application` everywhere it's needed
+* We are unconstrained in what a function can do - if it receives application,
+  it can do anything contained within that interface
+
+## Getting rid of `IO`
+If we add a generic type parameter to the handles, we can abstract away `IO`:
+
+```haskell
+data Persistence m =
+  { persistUser :: UserName -> Password -> m UserId
+  , getUser :: UserId -> m (Maybe User)
+  }
+
+data Application m =
+  { persistence :: Persistence m
+  , validateUsername :: UserId -> m Bool
+  }
+```
+
+Now, if we wanted to - we can actually run these as pure functions by using the
+`Identity` monad.
```

Comment:
You can even use `type family Id t where Id t = t` if you don't want to deal with `Identity` :wink: